test/src/sdk-harness/test_projects/storage_map/src/main.sw

```diff
@@ -1,37 +1,72 @@
 contract;
 
+use std::hash::*;
+
 pub struct Struct {
     x: u32,
     y: b256,
     z: b256,
 }
 
+impl Hash for Struct {
+    fn hash(self, ref mut state: Hasher) {
+        self.x.hash(state);
+        self.y.hash(state);
+        self.z.hash(state);
+    }
+}
+
 pub enum Enum {
     V1: b256,
     V2: u64,
     V3: b256,
 }
 
+impl Hash for Enum {
+    fn hash(self, ref mut state: Hasher) {
+        match self {
+            Enum::V1(val) => {
+                0_u8.hash(state);
+                val.hash(state);
+            }
+            Enum::V2(val) => {
+                1_u8.hash(state);
+                val.hash(state);
+            }
+            Enum::V3(val) => {
+                2_u8.hash(state);
+                val.hash(state);
```

Comment:
I'm wondering if introducing a compiler intrinsic to get the number of an `enum` tag (like those `0_u8`, `1_u8`, etc) would be useful. In principle, we should have a compiler synthesized hashing function for structs, enums, etc, but still something like [std](https://doc.rust-lang.org/std/index.html)::[mem](https://doc.rust-lang.org/std/mem/index.html)::[discriminant](https://doc.rust-lang.org/std/mem/fn.discriminant.html#) should be useful.

Reply:
This would simplify defining enum hashes a bit but on the other end, we want maximum stability on this. Changing the enum definition should not change the hash implicitly.

There is a section in std::mem::discriminant that says : 
```The discriminant of an enum variant may change if the enum definition changes.```

Not sure there is a way of generating the same discriminant when the enum variants are added in between and when variants are renamed.

I would keep this explicit because it is more stable.